src/lib.rs

```diff
@@ -438,12 +438,25 @@ pub trait FallibleIterator {
     ///
     /// An `Err` will be returned if any invocation of `next` returns `Err`.
     #[inline]
-    fn collect<T>(self) -> Result<T, Self::Error>
+    fn collect<T>(mut self) -> Result<T, Self::Error>
     where
-        T: FromFallibleIterator<Self::Item>,
+        T: std::iter::FromIterator<Self::Item>,
         Self: Sized,
     {
-        T::from_fallible_iter(self)
+        let mut first_err = None;
+        let res = T::from_iter(std::iter::from_fn(|| {
```

Comment:
You should just be able to do `self.iterator().collect()`, since `Result<T, E>` implements `FromIterator` from iterators over results.